tests/storage/databases/test_state_store.py

```diff
@@ -0,0 +1,109 @@
+from typing import Dict, List, Sequence, Tuple
+from unittest.mock import patch
+
+from twisted.internet.defer import Deferred, ensureDeferred
+
+from synapse.storage.state import StateFilter
+from synapse.types import MutableStateMap, StateMap
+
+from tests.unittest import HomeserverTestCase
+
+
+class StateGroupInflightCachingTestCase(HomeserverTestCase):
+    def setUp(self) -> None:
+        super(StateGroupInflightCachingTestCase, self).setUp()
+        # Patch out the `_get_state_groups_from_groups`.
+        # This is useful because it lets us pretend we have a slow database.
+        gsgfg_patch = patch(
+            "synapse.storage.databases.state.store.StateGroupDataStore._get_state_groups_from_groups",
+            self._fake_get_state_groups_from_groups,
+        )
+        gsgfg_patch.start()
+        self.addCleanup(gsgfg_patch.stop)
+        self.gsgfg_calls: List[
+            Tuple[Tuple[int, ...], StateFilter, Deferred[Dict[int, StateMap[str]]]]
+        ] = []
+
+    def prepare(self, reactor, clock, homeserver) -> None:
+        super(StateGroupInflightCachingTestCase, self).prepare(
+            reactor, clock, homeserver
+        )
+        self.state_storage = homeserver.get_storage().state
+        self.state_datastore = homeserver.get_datastores().state
+
+    def _fake_get_state_groups_from_groups(
```

Comment:
This looks like it could be replaced by a `side_effect=Deferred` when mocking the call and then using the standard methods on `Mock` to get back the arguments / results of calls...but maybe not.

Reply:
I can't see how to get the results off a `Mock` — if you know how that would be appreciated!

Reply:
[`call_args` (or `call_args_list`)](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.call_args) should have it.

Reply:
I've tried my best to read that carefully multiple times and I don't see it make any mention of the return values, just the `*args` and `**kwargs`. Am I just being blind?

Reply:
Oh, yes I don't think there's a way to get the return value of the calls. Sorry, misread!